File: api/review_service.py
"""Review queue service layer for HITL workflow."""

from __future__ import annotations

import logging
from datetime import UTC, datetime
from typing import TYPE_CHECKING

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from models import Document, DocumentStatus, ReviewAction, ReviewActionType, User

if TYPE_CHECKING:
    from collections.abc import Sequence

logger = logging.getLogger(__name__)


async def get_review_queue(
    session: AsyncSession,
    status: DocumentStatus | None = None,
    limit: int = 50,
    offset: int = 0,
) -> Sequence[Document]:
    """
    Get documents in review queue.

    Args:
        session: Database session
        status: Filter by document status (default: queued documents)
        limit: Maximum number of documents to return
        offset: Number of documents to skip

    Returns:
        List of documents with their text content loaded
    """
    query = select(Document).options(selectinload(Document.text_content))

    # Default to queued documents if no status specified
    if status is None:
        query = query.where(Document.status == DocumentStatus.QUEUED)
    else:
        query = query.where(Document.status == status)

    # Order by upload time (oldest first for fairness)
    query = query.order_by(Document.uploaded_at.asc())
    query = query.limit(limit).offset(offset)

    result = await session.execute(query)
    return result.scalars().all()


async def claim_document(
    session: AsyncSession,
    document_id: int,
    reviewer: User,
) -> Document:
    """
    Claim a document for review.

    Args:
        session: Database session
        document_id: Document ID to claim
        reviewer: User claiming the document

    Returns:
        Updated document

    Raises:
        ValueError: If document not found, already claimed, or not in queued status
    """
    # Load document
    result = await session.execute(select(Document).where(Document.id == document_id))
    document = result.scalar_one_or_none()

    if not document:
        msg = f"Документ {document_id} не найден"
        raise ValueError(msg)

    if document.status != DocumentStatus.QUEUED:
        msg = f"Документ {document_id} не может быть принят (статус: {document.status.value})"
        raise ValueError(msg)

    if document.assigned_reviewer_id is not None:
        msg = f"Документ {document_id} уже принят другим рецензентом"
        raise ValueError(msg)

    # Claim document
    document.assigned_reviewer_id = reviewer.id
    document.status = DocumentStatus.IN_REVIEW
    document.updated_at = datetime.now(UTC)

    # Create review action
    action = ReviewAction(
        document_id=document.id,
        reviewer_id=reviewer.id,
        action=ReviewActionType.CLAIM,
        from_category=document.category_predicted,
        to_category=None,
        comment=None,
        duration_seconds=None,
    )
    session.add(action)

    await session.commit()
    await session.refresh(document)

    logger.info(
        f"Документ {document_id} принят рецензентом {reviewer.id} ({reviewer.email})"
    )

    return document


async def release_document(
    session: AsyncSession,
    document_id: int,
    reviewer: User,
) -> Document:
    """
    Release a claimed document back to queue.

    Args:
        session: Database session
        document_id: Document ID to release
        reviewer: User releasing the document

    Returns:
        Updated document

    Raises:
        ValueError: If document not found, not claimed by this reviewer
    """
    # Load document
    result = await session.execute(select(Document).where(Document.id == document_id))
    document = result.scalar_one_or_none()

    if not document:
        msg = f"Документ {document_id} не найден"
        raise ValueError(msg)

    if document.status != DocumentStatus.IN_REVIEW:
        msg = (
            f"Документ {document_id} не находится в обработке (статус: {document.status.value})"
        )
        raise ValueError(msg)

    if document.assigned_reviewer_id != reviewer.id:
        msg = f"Документ {document_id} не закреплён за этим рецензентом"
        raise ValueError(msg)

    # Release document
    document.assigned_reviewer_id = None
    document.status = DocumentStatus.QUEUED
    document.updated_at = datetime.now(UTC)

    # Create review action
    action = ReviewAction(
        document_id=document.id,
        reviewer_id=reviewer.id,
        action=ReviewActionType.RELEASE,
        from_category=document.category_predicted,
        to_category=None,
        comment=None,
        duration_seconds=None,
    )
    session.add(action)

    await session.commit()
    await session.refresh(document)

    logger.info(
        f"Документ {document_id} возвращён в очередь рецензентом {reviewer.id} ({reviewer.email})"
    )

    return document


async def resolve_document(  # noqa: PLR0913
    session: AsyncSession,
    document_id: int,
    reviewer: User,
    final_category: str,
    applicant_name: str | None = None,
    applicant_lastname: str | None = None,
    comment: str | None = None,
) -> Document:
    """
    Resolve a document review.

    Args:
        session: Database session
        document_id: Document ID to resolve
        reviewer: User resolving the document
        final_category: Final category after review
        applicant_name: Updated applicant name (optional)
        applicant_lastname: Updated applicant lastname (optional)
        comment: Reviewer notes (optional)

    Returns:
        Updated document

    Raises:
        ValueError: If document not found, not claimed by this reviewer
    """
    # Load document and calculate duration
    result = await session.execute(select(Document).where(Document.id == document_id))
    document = result.scalar_one_or_none()

    if not document:
        msg = f"Документ {document_id} не найден"
        raise ValueError(msg)

    if document.status != DocumentStatus.IN_REVIEW:
        msg = (
            f"Документ {document_id} не находится в обработке (статус: {document.status.value})"
        )
        raise ValueError(msg)

    if document.assigned_reviewer_id != reviewer.id:
        msg = f"Документ {document_id} не закреплён за этим рецензентом"
        raise ValueError(msg)

    # Find the claim action to calculate duration
    claim_result = await session.execute(
        select(ReviewAction)
        .where(
            ReviewAction.document_id == document_id,
            ReviewAction.reviewer_id == reviewer.id,
            ReviewAction.action == ReviewActionType.CLAIM,
        )
        .order_by(ReviewAction.created_at.desc())
        .limit(1)
    )
    claim_action = claim_result.scalar_one_or_none()

    duration_seconds = None
    if claim_action:
        duration = datetime.now(UTC) - claim_action.created_at
        duration_seconds = int(duration.total_seconds())

    # Determine action type
    action_type = (
        ReviewActionType.ACCEPT
        if final_category == document.category_predicted
        else ReviewActionType.OVERRIDE
    )

    # Update document
    document.category_final = final_category
    document.status = DocumentStatus.RESOLVED
    document.updated_at = datetime.now(UTC)

    if applicant_name is not None:
        document.applicant_name = applicant_name
    if applicant_lastname is not None:
        document.applicant_lastname = applicant_lastname

    # Create review action
    action = ReviewAction(
        document_id=document.id,
        reviewer_id=reviewer.id,
        action=action_type,
        from_category=document.category_predicted,
        to_category=final_category,
        comment=comment,
        duration_seconds=duration_seconds,
    )
    session.add(action)

    await session.commit()
    await session.refresh(document)

    logger.info(
        f"Документ {document_id} обработан рецензентом {reviewer.id} "
        f"({reviewer.email}): {document.category_predicted} -> {final_category} "
        f"за {duration_seconds}s"
    )

    return document


async def get_document_audit_trail(
    session: AsyncSession,
    document_id: int,
) -> Sequence[ReviewAction]:
    """
    Get audit trail for a document.

    Args:
        session: Database session
        document_id: Document ID

    Returns:
        List of review actions ordered by creation time
    """
    result = await session.execute(
        select(ReviewAction)
        .options(selectinload(ReviewAction.reviewer))
        .where(ReviewAction.document_id == document_id)
        .order_by(ReviewAction.created_at.asc())
    )
    return result.scalars().all()


async def get_document_by_id(
    session: AsyncSession,
    document_id: int,
) -> Document | None:
    """
    Get a document by ID with related data loaded.

    Args:
        session: Database session
        document_id: Document ID

    Returns:
        Document with text content and assigned reviewer loaded, or None
    """
    result = await session.execute(
        select(Document)
        .options(
            selectinload(Document.text_content),
            selectinload(Document.assigned_reviewer),
        )
        .where(Document.id == document_id)
    )
    return result.scalar_one_or_none()


__all__ = [
    "claim_document",
    "get_document_audit_trail",
    "get_document_by_id",
    "get_review_queue",
    "release_document",
    "resolve_document",
]

File: api/app/Services/ReviewService.php
<?php

namespace App\Services;

use App\Models\Document;
use App\Models\ReviewAction;
use App\Models\User;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class ReviewService
{
    /**
     * Get documents in the review queue.
     *
     * Mirrors get_review_queue() from Python.
     */
    public function getQueue(?string $status, int $limit, ?string $cursor): array
    {
        $query = Document::with(['text', 'reviewer'])
            ->whereNull('processing_state');

        if ($status !== null) {
            $query->where('status', $status);
        }

        $cursorData = $this->decodeCursor($cursor);
        if ($cursorData !== null) {
            $query->where(function ($builder) use ($cursorData) {
                $builder->where('created_at', '>', $cursorData['created_at'])
                    ->orWhere(function ($inner) use ($cursorData) {
                        $inner->where('created_at', '=', $cursorData['created_at'])
                            ->where('id', '>', $cursorData['id']);
                    });
            });
        }

        $documents = $query
            ->orderBy('created_at', 'asc')
            ->orderBy('id', 'asc')
            ->limit($limit + 1)
            ->get();

        $nextCursor = null;
        if ($documents->count() > $limit) {
            $last = $documents->pop();
            if ($last) {
                $nextCursor = $this->encodeCursor(
                    (string) $last->created_at,
                    (string) $last->id,
                );
            }
        }

        return [
            'documents' => $documents,
            'next_cursor' => $nextCursor,
        ];
    }

    /**
     * Claim a document for review.
     *
     * Mirrors claim_document() — uses a DB transaction to prevent double-claim.
     */
    public function claim(string $documentId, User $reviewer): Document
    {
        return DB::transaction(function () use ($documentId, $reviewer) {
            /** @var Document|null $document */
            $document = Document::with('text')->lockForUpdate()->find($documentId);

            if (! $document) {
                throw new \InvalidArgumentException("Документ {$documentId} не найден");
            }

            if ($document->status !== 'queued') {
                throw new \InvalidArgumentException(
                    "Документ {$documentId} не может быть принят (статус: {$document->status})"
                );
            }

            if ($document->assigned_reviewer_id !== null) {
                throw new \InvalidArgumentException(
                    "Документ {$documentId} уже принят другим рецензентом"
                );
            }

            $now = now();
            $document->update([
                'assigned_reviewer_id' => $reviewer->id,
                'status'               => 'in_review',
                'review_started_at'    => $now,
                'updated_at'           => $now,
            ]);

            ReviewAction::create([
                'document_id'   => $documentId,
                'reviewer_id'   => $reviewer->id,
                'action'        => 'claim',
                'from_category' => $document->category_predicted,
                'to_category'   => null,
                'comment'       => null,
                'duration_seconds' => null,
                'created_at'    => $now,
            ]);

            Log::info("Document {$documentId} claimed by reviewer {$reviewer->email}");

            return $document->fresh(['text', 'reviewer']);
        });
    }

    /**
     * Release a claimed document back to queue.
     *
     * Mirrors release_document() from Python.
     */
    public function release(string $documentId, User $reviewer): Document
    {
        return DB::transaction(function () use ($documentId, $reviewer) {
            /** @var Document|null $document */
            $document = Document::with('text')->lockForUpdate()->find($documentId);

            if (! $document) {
                throw new \InvalidArgumentException("Документ {$documentId} не найден");
            }

            if ($document->status !== 'in_review') {
                throw new \InvalidArgumentException(
                    "Документ {$documentId} не находится в обработке (статус: {$document->status})"
                );
            }

            if ($document->assigned_reviewer_id !== $reviewer->id) {
                throw new \InvalidArgumentException(
                    "Документ {$documentId} не закреплён за этим рецензентом"
                );
            }

            $now = now();
            $document->update([
                'assigned_reviewer_id' => null,
                'status'               => 'queued',
                'review_started_at'    => null,
                'updated_at'           => $now,
            ]);

            ReviewAction::create([
                'document_id'      => $documentId,
                'reviewer_id'      => $reviewer->id,
                'action'           => 'release',
                'from_category'    => $document->category_predicted,
                'to_category'      => null,
                'comment'          => null,
                'duration_seconds' => null,
                'created_at'       => $now,
            ]);

            Log::info("Document {$documentId} released by reviewer {$reviewer->email}");

            return $document->fresh(['text', 'reviewer']);
        });
    }

    /**
     * Resolve a document review with a final category.
     *
     * Mirrors resolve_document() from Python.
     */
    public function resolve(
        string $documentId,
        User $reviewer,
        string $finalCategory,
        ?string $applicantName,
        ?string $applicantLastname,
        ?string $comment
    ): Document {
        return DB::transaction(function () use ($documentId, $reviewer, $finalCategory, $applicantName, $applicantLastname, $comment) {
            /** @var Document|null $document */
            $document = Document::with('text')->lockForUpdate()->find($documentId);

            if (! $document) {
                throw new \InvalidArgumentException("Документ {$documentId} не найден");
            }

            if ($document->status !== 'in_review') {
                throw new \InvalidArgumentException(
                    "Документ {$documentId} не находится в обработке (статус: {$document->status})"
                );
            }

            if ($document->assigned_reviewer_id !== $reviewer->id) {
                throw new \InvalidArgumentException(
                    "Документ {$documentId} не закреплён за этим рецензентом"
                );
            }

            $durationSeconds = null;
            if ($document->review_started_at) {
                $durationSeconds = (int) $document->review_started_at->diffInSeconds(now());
            }

            // Determine action type
            $actionType = $finalCategory === $document->category_predicted
                ? 'accept'
                : 'override';

            $now = now();
            $updates = [
                'category_final' => $finalCategory,
                'status'         => 'resolved',
                'resolved_at'    => $now,
                'updated_at'     => $now,
            ];
            if ($applicantName !== null) {
                $updates['applicant_name'] = $applicantName;
                $updates['applicant_name_normalized'] = $this->normalizeApplicantLookup($applicantName);
            }
            if ($applicantLastname !== null) {
                $updates['applicant_lastname'] = $applicantLastname;
                $updates['applicant_lastname_normalized'] = $this->normalizeApplicantLookup($applicantLastname);
            }
            $document->update($updates);

            ReviewAction::create([
                'document_id'      => $documentId,
                'reviewer_id'      => $reviewer->id,
                'action'           => $actionType,
                'from_category'    => $document->category_predicted,
                'to_category'      => $finalCategory,
                'comment'          => $comment,
                'duration_seconds' => $durationSeconds,
                'created_at'       => $now,
            ]);

            Log::info(
                "Document {$documentId} resolved by {$reviewer->email}: "
                ."{$document->category_predicted} -> {$finalCategory} in {$durationSeconds}s"
            );

            return $document->fresh(['text', 'reviewer']);
        });
    }

    /**
     * Get the audit trail for a document.
     *
     * Mirrors get_document_audit_trail() from Python.
     */
    public function getAuditTrail(string $documentId): \Illuminate\Database\Eloquent\Collection
    {
        return ReviewAction::with('reviewer')
            ->where('document_id', $documentId)
            ->orderBy('created_at', 'asc')
            ->get();
    }

    /**
     * Get a single document with related data.
     */
    public function findDocument(string $documentId): ?Document
    {
        return Document::with(['text', 'reviewer'])->find($documentId);
    }

    private function encodeCursor(string $createdAt, string $id): string
    {
        return rtrim(strtr(base64_encode(json_encode([
            'created_at' => $createdAt,
            'id' => $id,
        ], JSON_THROW_ON_ERROR)), '+/', '-_'), '=');
    }

    private function decodeCursor(?string $cursor): ?array
    {
        if (! $cursor) {
            return null;
        }

        $decoded = base64_decode(strtr($cursor, '-_', '+/'), true);
        if ($decoded === false) {
            return null;
        }

        try {
            $payload = json_decode($decoded, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return null;
        }

        if (! isset($payload['created_at'], $payload['id'])) {
            return null;
        }

        return [
            'created_at' => $payload['created_at'],
            'id' => $payload['id'],
        ];
    }

    private function normalizeApplicantLookup(string $value): string
    {
        $normalized = mb_strtolower(trim($value));
        $normalized = preg_replace('/[^\p{L}\p{N}]+/u', ' ', $normalized) ?? '';

        return trim($normalized);
    }
}
